contracts/AcceleratingDistributorClaimAndStake.sol

```diff
@@ -0,0 +1,81 @@
+
+/**
+ * @notice Across token distribution contract. Contract is inspired by Synthetix staking contract and Ampleforth geyser.
+ * Stakers start by earning their pro-rata share of a baseEmissionRate per second which increases based on how long
+ * they have staked in the contract, up to a max emission rate of baseEmissionRate * maxMultiplier. Multiple LP tokens
+ * can be staked in this contract enabling depositors to batch stake and claim via multicall. Note that this contract is
+ * only compatible with standard ERC20 tokens, and not tokens that charge fees on transfers, dynamically change
+ * balance, or have double entry-points. It's up to the contract owner to ensure they only add supported tokens.
+ */
+
+contract AcceleratingDistributorClaimAndStake is AcceleratingDistributor {
+    // Contract which rewards tokens to users that they can then stake. MerkleDistributor logic does not impact
+    // this contract at all, but its stored here for convenience to allow claimAndStake to be called by a user to
+    // claim their staking tokens and stake atomically.
+    MerkleDistributorInterface public merkleDistributor;
+
+    /**************************************
+     *               EVENTS               *
+     **************************************/
+
+    event SetMerkleDistributor(address indexed newMerkleDistributor);
+
+    constructor(address _rewardToken) AcceleratingDistributor(_rewardToken) {}
+
+    /**************************************
+     *          ADMIN FUNCTIONS           *
+     **************************************/
+
+    /**
+     * @notice Resets merkle distributor contract called in claimAndStake()
+     * @param _merkleDistributor Address to set merkleDistributor to.
+     */
+    function setMerkleDistributor(address _merkleDistributor) external onlyOwner {
+        merkleDistributor = MerkleDistributorInterface(_merkleDistributor);
+        emit SetMerkleDistributor(_merkleDistributor);
+    }
+
+    /**************************************
+     *          STAKER FUNCTIONS          *
+     **************************************/
+
+    /**
+     * @notice Claim tokens from a MerkleDistributor contract and stake them for rewards.
+     * @dev Will revert if `merkleDistributor` is not set to valid MerkleDistributor contract.
+     * @dev Will revert if any of the claims recipient accounts are not equal to caller, or if any reward token
+     *      for claim is not a valid staking token or are not the same token as the other claims.
+     * @dev Will revert if this contract is not a "whitelisted claimer" on the MerkleDistributor contract.
+     * @dev The caller of this function must approve this contract to spend total amount of stakedToken.
+     * @param claims Claim leaves to retrieve from MerkleDistributor.
+     * @param stakedToken The address of the token to stake.
+     */
+    function claimAndStake(MerkleDistributorInterface.Claim[] memory claims, address stakedToken)
```

Comment:
I don't want to throw a wrench into anything, but I was brainstorming a bit and wanted to ask about a slightly different implementation/strategy.

This would introduce less code in this contract, but may add more code (or more complex code) in the merkle distributor and may have a slightly better business outcome (one transaction, no approval):
- Add a stakeFor() method in this contract (basically just a wrapper function around _stake() that you have above).
- Drop the claimAndStake() method.
- Add a claimFor method in the merkle distributor that allows a contract to claim on behalf of a user (and actually receive the tokens). This could be done via signature or whitelist.
- Build a special intermediary contract that does the following:
  - This contract would have a single method to do 2 things:
    - Claim their airdrop (and pull it into this contract).
    - Call stakeFor() to stake on behalf of the user.

This process has a few advantages:
1. Lighter modifications to the AcceleratingDistributor (still some, though), which holds user funds.
2. Single transaction claim (no approvals).

Downsides:
1. More cross contract calls.
2. One more contract to deploy.

Reply:
I think this is a good idea I'll start work on the AcrossMerkleDistributor